rust/otap-dataflow/crates/otap/src/batch_processor.rs

```diff
@@ -542,7 +683,154 @@ impl local::Processor<OtapPdata> for BatchProcessor {
     }
 }
 
+impl SignalBatches {
+    fn new(config: &Config) -> Self {
+        Self {
+            logs: SignalBuffer::new(config),
+            traces: SignalBuffer::new(config),
+            metrics: SignalBuffer::new(config),
+        }
+    }
+}
+
+impl BatchPortion {
+    fn new(inkey: Option<SlotKey>, items: usize) -> Self {
+        Self { inkey, items }
+    }
+}
+
+impl MultiContext {
+    fn new(inputs: Vec<BatchPortion>) -> Self {
+        Self { inputs, pos: 0 }
+    }
+}
+
+impl Inputs {
+    fn drain(&mut self) -> Self {
+        Self {
+            pending: self.pending.drain(..).collect(),
+            context: self.context.drain(..).collect(),
+            items: std::mem::take(&mut self.items),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.items == 0
+    }
+
+    fn requests(&self) -> usize {
+        self.pending.len()
+    }
+
+    fn accept(&mut self, batch: OtapArrowRecords, part: BatchPortion) {
+        self.items += part.items;
+        self.pending.push(batch);
+        self.context.push(part);
+    }
+
+    fn take_pending(&mut self) -> Vec<OtapArrowRecords> {
+        std::mem::take(&mut self.pending)
+    }
+
+    fn take_context(&mut self) -> MultiContext {
+        MultiContext::new(std::mem::take(&mut self.context))
+    }
+}
+
 impl SignalBuffer {
```

Comment:
nit: it might be nice to have some brief comments about what the new methods we've implemented are doing (`take_remaining`, `drain_context` and `handle_partial_responses`).